### PR TITLE
[Compile Cache] Dont pass None args to Compile Cache

### PR DESCRIPTION
**Pull request by anijain2305**
@Chillee What do you think here? This is little ugly but better than doing it in C++.

~~Basically, we send only the non-None args to the cache. Our cache supports scenario where same function can have different number of arguments. So, it recompiles when it all args are Tensor the next time.~~

Update - I changed this to C++ because, as Horace pointed out, it does not work in all cases.

### CODE REVIEW

**Comment by Chillee** [comment]
What happens if the user has something like
```
f(None, a)
f(a, None)
```
?

More generally, I think this logic is starting to make unclear what invariants we're expecting at each layer. It's probably worth documenting those.

**Comment by anijain2305** [comment]
> What happens if the user has something like
> 
> ```
> f(None, a)
> f(a, None)
> ```
> 
> ?
> 
> More generally, I'm worried this logic is starting to make unclear what invariants we're expecting at each layer. It's probably worth documenting those (perhaps with type hints).

You are right. Your example would cause failure here.

**Comment by hyunwoongko** [comment]
I am trying to train the HuggingFace model using AOTAutograd. So I am interested in this issue. What if I pass `fn(input_ids, None, attention_mask)` to a function like `fn(input_ids, layer_past, attention_mask)` ?

**Comment by anijain2305** [comment]
> I am trying to train the HuggingFace model using AOTAutograd. So I am interested in this issue. What if I pass `(input_ids, None, attention_mask)` or `(input_ids, layer_past, None)` to a function like `fn(input_ids, layer_past, attention_mask)` ? Doesn't AOTAutograd also allow non tensor inputs like `torch.jit.trace`?

Thanks Kevin. Yes, your scenario also falls in the same category that this PR is trying to solve. 

Yes, currently there is a bug that we can't handle None args. I will work on it to get it resolved. If you have a small unit test case, I will be happy to add it to our harness.

File: test/test_compile_cache.py [hunk 1]
```diff
@@ -608,6 +608,62 @@ def fn(x, tuple_static_arg):
...
+        functorch.compile.clear_compile_cache()
+
+        start_num_recomps = functorch.compile.num_of_recompilations()
+
+        aot_autograd_f = aot_function(fn, nop, nop)
+
+        a = torch.randn(2, 2, requires_grad=True)
+        b = torch.randn(2, 2, requires_grad=True)
+        c = None
+        check(a, b, c, aot_autograd_f, fn)
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply test/test_compile_cache.py]: Done.

**Comment by anijain2305** [comment]
@Chillee This is ready. CI failures are unrelated.

**Review by Chillee — APPROVED** [review]
Could you measure the overhead and post it before you land?

**Comment by anijain2305** [comment]
> Could you measure the overhead and post it before you land?

Oh yes, I measured that.

For a function with 3 args, cache lookup latency is as follows
This PR : 1.33 us
Main branch : 1.29 us

I didn't see any noticeable change in latency.

**Comment by anijain2305** [comment]
@hyunwoongko You can try your test case with this PR.

**Comment by hyunwoongko** [comment]
It's great! Thank you so much

**Pull request merged**